Mathlib/Data/Nat/Factorization/Root.lean

```diff
@@ -0,0 +1,99 @@
+variable {a b n : ℕ}
+
+/-- Flooring root of a natural number. This divides the valuation of every prime number rounding
+down.
+
+Eg if `n = 2`, `a = 2^3 * 3^2 * 5`, then `floorRoot n a = 2 * 3`.
+
+In order theory terms, this is the upper or right adjoint of the map `a ↦ a ^ n`. -/
+noncomputable def floorRoot (n a : ℕ) : ℕ := (a.factorization ⌊/⌋ n).prod (· ^ ·)
+
+@[simp] lemma floorRoot_zero_left (a : ℕ) : floorRoot 0 a = 1 := by simp [floorRoot]
+@[simp] lemma floorRoot_zero_right (n : ℕ) : floorRoot n 0 = 1 := by simp [floorRoot]
+@[simp] lemma floorRoot_one_left (ha : a ≠ 0) : floorRoot 1 a = a := by simp [floorRoot, ha]
+@[simp] lemma floorRoot_one_right (n : ℕ) : floorRoot n 1 = 1 := by simp [floorRoot]
+
+@[simp] lemma floorRoot_pow_self (hn : n ≠ 0) (ha : a ≠ 0) : floorRoot n (a ^ n) = a := by
+  simp [floorRoot, pos_iff_ne_zero.2 hn, ha]
+
+@[simp] lemma floorRoot_ne_zero : floorRoot n a ≠ 0 := by
+  simp (config := { contextual := true }) [floorRoot, not_imp_not]
+
+@[simp] lemma factorization_floorRoot (n a : ℕ) :
+    (floorRoot n a).factorization = a.factorization ⌊/⌋ n := by
+  refine prod_pow_factorization_eq_self fun p hp ↦ ?_
+  have : p.Prime ∧ _ := by simpa using support_floorDiv_subset hp
+  exact this.1
+
+lemma pow_dvd_iff (hn : n ≠ 0) (ha : a ≠ 0) (hb : b ≠ 0) : a ^ n ∣ b ↔ a ∣ floorRoot n b := by
+  rw [← factorization_le_iff_dvd (pow_ne_zero _ ha) hb,
+    ← factorization_le_iff_dvd ha floorRoot_ne_zero, factorization_pow, factorization_floorRoot,
+    le_floorDiv_iff_smul_le (β := ℕ →₀ ℕ) (pos_iff_ne_zero.2 hn)]
+
+lemma floorRoot_pow_dvd (hn : n ≠ 0) (ha : a ≠ 0) : floorRoot n a ^ n ∣ a :=
+  (pow_dvd_iff hn floorRoot_ne_zero ha).2 dvd_rfl
+
+/-- Ceiling root of a natural number. This divides the valuation of every prime number rounding up.
+
+Eg if `n = 3`, `a = 2^4 * 3^2 * 5`, then `ceilRoot n a = 2^2 * 3 * 5`.
+
+In order theory terms, this is the lower or left adjoint of the map `a ↦ a ^ n`. -/
+noncomputable def ceilRoot (n a : ℕ) : ℕ := (a.factorization ⌈/⌉ n).prod (· ^ ·)
+
+@[simp] lemma ceilRoot_zero_left (a : ℕ) : ceilRoot 0 a = 1 := by simp [ceilRoot]
+@[simp] lemma ceilRoot_zero_right (n : ℕ) : ceilRoot n 0 = 1 := by simp [ceilRoot]
+@[simp] lemma ceilRoot_one_left (ha : a ≠ 0) : ceilRoot 1 a = a := by simp [ceilRoot, ha]
+@[simp] lemma ceilRoot_one_right (n : ℕ) : ceilRoot n 1 = 1 := by simp [ceilRoot]
+
+@[simp] lemma ceilRoot_pow_self (hn : n ≠ 0) (ha : a ≠ 0) : ceilRoot n (a ^ n) = a := by
+  simp [ceilRoot, pos_iff_ne_zero.2 hn, ha]
+
+@[simp] lemma ceilRoot_ne_zero : ceilRoot n a ≠ 0 := by
+  simp (config := { contextual := true }) [ceilRoot, not_imp_not]
+
+@[simp] lemma factorization_ceilRoot (n a : ℕ) :
+    (ceilRoot n a).factorization = a.factorization ⌈/⌉ n := by
+  refine prod_pow_factorization_eq_self fun p hp ↦ ?_
+  have : p.Prime ∧ _ := by simpa using support_ceilDiv_subset hp
+  exact this.1
+
+lemma dvd_pow_iff (hn : n ≠ 0) (ha : a ≠ 0) : a ∣ b ^ n ↔ ceilRoot n a ∣ b := by
```

Comment:
It looks like you need `b ≠ 0 ∨ n ≠ 0` here.

Reply:
This can't possibly work. When plugging `n = 0`, the statement reduces to `a = 1 ↔ ceilRoot 0 a ∣ b`, which can be true for all `b` only if `a = 1` and `ceilRoot 0 1 = 1`.

For the record:
* When plugging `b = 0`, the statement reduces to `n = 0 → a = 1`, which is okay if we assume `n ≠ 0`.
* When plugging `a = 0`, the statement reduces to `b = 0 ∧ n ≠ 0 ↔ ceilRoot n 0 ∣ b`, which is okay if we assume `n ≠ 0` and set `ceilRoot n 0 = 0`.

Reply:
It would be possible to make the implication `a ∣ b ^ n → ceilRoot n a ∣ b` unconditional by setting `ceilRoot 0 a = 0` if `a ≠ 0`. Do you think that's worth the asymmetry with `floorRoot`?